### PR TITLE
Complete the initial third-party API implementation in the master.

### PR DESCRIPTION
**Pull request by brendandburns**
Depends on https://github.com/kubernetes/kubernetes/pull/13765

@thockin @caesarxuchao @lavalamp this is part of implementing

General design doc for this feature is here:

https://github.com/kubernetes/kubernetes/blob/master/docs/design/extending-api.md

Original design thread/review/rfc:

https://github.com/kubernetes/kubernetes/pull/11781

### CODE REVIEW

**Comment by brendandburns** [comment]
@caesarxuchao @lavalamp this is now ready for review.  Please only review https://github.com/kubernetes/kubernetes/commit/9f9f74cb76150be7eec7f9e4d850487dcd463917 as the rest of this PR is in #13765

Thanks!
--brendan

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by brendandburns** [comment]
@caesarxuchao all green and ready for review ;)

**Comment by caesarxuchao** [comment]
LGTM except a few nits. @lavalamp is having a quick pass now.

File: pkg/api/meta/restmapper.go [hunk 1]
```diff
@@ -171,14 +171,14 @@ func (m *DefaultRESTMapper) ResourceSingularizer(resource string) (singular stri
 func (m *DefaultRESTMapper) VersionAndKindForResource(resource string) (defaultVersion, kind string, err error) {
 	meta, ok := m.mapping[strings.ToLower(resource)]
 	if !ok {
-		return "", "", fmt.Errorf("no resource %q has been defined", resource)
+		return "", "", fmt.Errorf("in version and kind for resource no resource %q has been defined", resource)
```
Comment (lavalamp) [inline pkg/api/meta/restmapper.go]: print out something identifying m, like m.group? Hm, I guess that's not public, so maybe not.

File: pkg/master/master.go [hunk 2]
```diff
@@ -774,7 +782,85 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
 	return version
 }
 
-func (m *Master) InstallThirdPartyAPI(rsrc *experimental.ThirdPartyResource) error {
+func (m *Master) HasThirdPartyResource(rsrc *expapi.ThirdPartyResource) (bool, error) {
```
Comment (lavalamp) [inline pkg/master/master.go]: Godoc

File: pkg/master/master.go [hunk 3]
```diff
@@ -774,7 +782,85 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
...
+	if found {
+		if err := m.removeAllThirdPartyResources(storage); err != nil {
+			return err
+		}
+		delete(m.thirdPartyResources, path)
+	}
+	return nil
+}
+
+func (m *Master) RemoveThirdPartyResource(path string) error {
```
Comment (lavalamp) [inline pkg/master/master.go]: Godoc

File: pkg/master/master.go [hunk 4]
```diff
@@ -774,7 +782,85 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
...
+	for ix := range list.Items {
+		item := &list.Items[ix]
+		if _, err := registry.Delete(ctx, item.Name, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Master) ListThirdPartyResources() []string {
```
Comment (lavalamp) [inline pkg/master/master.go]: :)

File: pkg/master/master.go [hunk 5]
```diff
@@ -242,6 +242,10 @@ type Master struct {
 
 	// storage for third party objects
 	thirdPartyStorage storage.Interface
+	// map from api path to storage for those objects
+	thirdPartyResources map[string]*thirdpartyresourcedataetcd.REST
+	// protects the map
+	thirdPartyResourcesLock sync.RWMutex
```
Comment (lavalamp) [inline pkg/master/master.go]: I guess it's a big change-- but I would prefer if the third party stuff was a separate object (and in a separate file) from the master.

**Comment by lavalamp** [comment]
Some thoughts; I don't think you need to block on them though.

File: pkg/api/meta/restmapper.go (hunk 1 above)
Comment (brendandburns) [inline pkg/api/meta/restmapper.go]: ack.  I don't think there's more we can output here.

File: pkg/master/master.go (hunk 2 above)
Comment (brendandburns) [inline pkg/master/master.go]: Done, though what is our policy on Godoc vis a vis interface implementations.  Shouldn't we defer to the comment in the interface specification?

File: pkg/master/master.go (hunk 3 above)
Comment (brendandburns) [inline pkg/master/master.go]: done, although really, isn't the name of the function sufficient?  (and also as above wrt to interfaces)

File: pkg/master/master.go (hunk 4 above)
Comment (brendandburns) [inline pkg/master/master.go]: done. although again, was there confusion about what this method did?  (and it implements an interface)

File: pkg/master/master.go (hunk 5 above)
Comment (brendandburns) [inline pkg/master/master.go]: If we did that (and I'm open to it, I suppose) you would still need to expose functions for mucking around with the go-restful web services that would look pretty similar to these functions.  I guess the other object could muck with `handlerContainer` directly, but that seems to break encapsulation boundaries.

**Comment by brendandburns** [comment]
Comments addressed, ptal.

Thanks!
--brendan

**Comment by caesarxuchao** [comment]
Shippable:

FAIL	k8s.io/kubernetes/pkg/master [build failed]

**Comment by caesarxuchao** [comment]
Thank you @brendandburns. I will apply the label once you fix the test.

File: pkg/master/master.go (hunk 2 above)
Comment (lavalamp) [inline pkg/master/master.go]: IMO the interface comment should describe the contract and the
implementation comment should describe what that means in specifically, but
I don't know that we have a rule.

File: pkg/master/master.go [hunk 6]
```diff
@@ -774,7 +782,89 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
...
+	if found {
+		if err := m.removeAllThirdPartyResources(storage); err != nil {
+			return err
+		}
+		delete(m.thirdPartyResources, path)
+	}
+	return nil
+}
+
+// RemoveThirdPartyResource removes all resources matching `path`.  Also deletes any stored data
```
Comment (lavalamp) [inline pkg/master/master.go]: Deletes from etcd?

File: pkg/master/master.go [hunk 7]
```diff
@@ -774,7 +782,89 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
...
+	for ix := range list.Items {
+		item := &list.Items[ix]
+		if _, err := registry.Delete(ctx, item.Name, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ListThirdPartyResources lists all currently installed third party resources
```
Comment (lavalamp) [inline pkg/master/master.go]: Installed means currently installed in the local master or installed in etcd?

Comment (lavalamp) [inline pkg/master/master.go]: (leaving these comments to illustrate what you can't figure out from the name of the function)

File: pkg/master/master.go [hunk 8]
```diff
@@ -774,7 +782,89 @@ func (m *Master) api_v1() *apiserver.APIGroupVersion {
...
+	return result
+}
+
+func (m *Master) addThirdPartyResourceStorage(path string, storage *thirdpartyresourcedataetcd.REST) {
+	m.thirdPartyResourcesLock.Lock()
+	defer m.thirdPartyResourcesLock.Unlock()
+	m.thirdPartyResources[path] = storage
+}
+
+// InstallThirdPartyResource installs a third party resource specified by 'rsrc'
```
Comment (lavalamp) [inline pkg/master/master.go]: Specific meaning of "installs" would be helpful.

File: pkg/master/master.go (hunk 6 above)
Comment (brendandburns) [inline pkg/master/master.go]: Implementation detail, totally irrelevant, it's noted that it deletes stored data, who cares where that data is stored, indeed the master has no idea where it is stored, it only has the storage interface.

File: pkg/master/master.go (hunk 7 above)
Comment (brendandburns) [inline pkg/master/master.go]: again, master knows nothing of etcd. so installed clearly means resources that the master/apiserver knows about and exports.

**Comment by caesarxuchao** [comment]
Add LGTM label.
Close and reopen to see shippable.

**Comment by dchen1107** [comment]
Removed lgtm from the pr since it requires a rebase.

**Comment by brendandburns** [comment]
Rebased.  Adding LGTM back in.

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Pull request merged**